components/codec/src/columnar/mod.rs

```diff
@@ -0,0 +1,550 @@
+mod float;
+mod int;
+mod timestamp;
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub(crate)))]
+pub enum Error {
+    #[snafu(display("Invalid version:{version}.\nBacktrace:\n{backtrace}"))]
+    InvalidVersion { version: u8, backtrace: Backtrace },
+
+    #[snafu(display("Invalid datum kind, err:{source}"))]
+    InvalidDatumKind { source: common_types::datum::Error },
+
+    #[snafu(display("No enough bytes to compose the nulls bit set.\nBacktrace:\n{backtrace}"))]
+    InvalidNullsBitSet { backtrace: Backtrace },
+
+    #[snafu(display(
+        "Datums is not enough, expect:{expect}, found:{found}.\nBacktrace:\n{backtrace}"
+    ))]
+    NotEnoughDatums {
+        expect: usize,
+        found: usize,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Failed to varint, err:{source}"))]
+    Varint { source: varint::Error },
+
+    #[snafu(display("Failed to do compact encoding, err:{source}"))]
+    CompactEncode { source: crate::compact::Error },
+}
+
+define_result!(Error);
+
+/// The trait bound on the encoders for different types.
+trait ValuesEncoder<T> {
+    /// Encode a batch of values into the `buf`.
+    ///
+    /// As the `estimated_encoded_size` method is provided, the `buf` should be
+    /// pre-allocate.
+    fn encode<B, I>(&self, buf: &mut B, values: I) -> Result<()>
+    where
+        B: BufMut,
+        I: Iterator<Item = T>;
+
+    /// The estimated size for memory pre-allocated.
+    fn estimated_encoded_size<I>(&self, values: I) -> usize
+    where
+        I: Iterator<Item = T>,
+    {
+        let (lower, higher) = values.size_hint();
+        let num = lower.max(higher.unwrap_or_default());
+        num * std::mem::size_of::<T>()
+    }
+}
+
+/// The trait bound on the decoders for different types.
+trait ValuesDecoder<T> {
+    fn decode<B, F>(&self, buf: &mut B, f: F) -> Result<()>
+    where
+        B: Buf,
+        F: FnMut(T) -> Result<()>;
+}
+
+/// The implementation for [`ValuesEncoder`].
+struct ValuesEncoderImpl;
+
+/// The implementation for [`ValuesDecoder`].
+struct ValuesDecoderImpl;
+
+#[derive(Clone, Debug)]
+pub struct ColumnarEncoder {
+    column_idx: u32,
+}
+
+/// A hint helps column encoding.
+pub struct EncodeHint {
+    pub num_nulls: Option<usize>,
+    pub num_datums: Option<usize>,
+    pub datum_kind: DatumKind,
+}
+
+impl EncodeHint {
+    fn compute_num_nulls<'a, I>(&mut self, datums: &I) -> usize
+    where
+        I: Iterator<Item = DatumView<'a>> + Clone,
+    {
+        if let Some(v) = self.num_nulls {
+            v
+        } else {
+            let num_nulls = datums.clone().filter(|v| v.is_null()).count();
+            self.num_nulls = Some(num_nulls);
+            num_nulls
+        }
+    }
+
+    fn compute_num_datums<'a, I>(&mut self, datums: &I) -> usize
+    where
+        I: Iterator<Item = DatumView<'a>> + Clone,
+    {
+        if let Some(v) = self.num_datums {
+            v
+        } else {
+            let num_datums = datums.clone().count();
```

Comment:
Invoke Iterator's size_hint?

Reply:
What we need is an exact count of the iterator, and I guess `size_hint` can't meet this requirement.